### PR TITLE
Rhmap 12487

### PR DESCRIPTION
**Pull request by pb82**
# Motivation

implements temp directory cleanup in fh-mbaas. There can be multiple mbaas instances and we need to make sure that the job runs on all nodes. This is ensured by using the number of instances (passed from supercore) and a `runList` which stores the nodes that have already completed the job.

TODO: tests

### CODE REVIEW

File: lib/handlers/api.js [hunk 1]
```diff
@@ -12,6 +12,8 @@ var eventMiddleware = require('../middleware/events');
 var appdataRouter = require('./app/data.js');
 var deleteEnvironmentData = require('../services/environment/deleteEnvironmentData.js');
 var logger = require('../util/logger.js').getLogger();
+var cleanup = require('../handlers/cleanup/tmp_cleanup');
```
Comment (rachael-oregan) [inline lib/handlers/api.js]: `const` can be used here

Comment (pb82) [inline lib/handlers/api.js]: The code base in fh-mbaas is largely still ES5 and so i decided to keep it that way. The alternative would be to move everything to const/let here.

File: lib/handlers/cleanup/tmp_cleanup.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Checks if a cleanup job already exists and returns its
+ * status
+ *
+ * @param req Http request
+ * @param callback Callback
+ */
+function runTmpCleanup(req, callback) {
+  CleanupStatus.find().then(status => {
+    if (!status || status.length === 0) {
```
Comment (rachael-oregan) [inline lib/handlers/cleanup/tmp_cleanup.js]: Could you do `!status.length` instead of `status.length === 0`

File: lib/jobs/tmp_cleanup_job.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ * @returns {boolean}
+ */
+function cleanupJobStale(job) {
+  let now = new Date().getTime();
+  let lastModified = now - job.modified.getTime();
+  return lastModified > (JOB_MAXAGE_SECONDS * 1000);
+}
+
+/**
+ * Try to delete all files and directories in the temp ddirectory.
```
Comment (rachael-oregan) [inline lib/jobs/tmp_cleanup_job.js]: Typo - `directory`

**Comment by pb82** [comment]
@wei-lee The tmp directory cleanup is actually not meant for the dyno: the idea is to clean the tmp directory of the mbaas nodes. The other two cleanup strategies (npm cache clean and fh-npm cleanup) are running on the dynos though.

Regarding the number of instances: there is a property called `feedhenry.numnodes` that i read using millicore-client.

**Comment by wei-lee** [comment]
@pb82 yeah, I saw that PR. I left a comment there.

My question really is why do we need to clean up the `tmp` dir on the mbaas nodes? Does it help free up more disk space of the environments for a customer?

Also, if that can help, any reason not doing it from fh-dynoman? If a user is performing cleanup one environment at a time, we don't need to clean both nodes - we just need to clean the `tmp` dir on the server that the environment is created on. If that is the case, wouldn't it make things a lot easier to just do it from fh-dynoman?

**Comment by pb82** [comment]
@wei-lee You're probably right. The ticket states: `To help manage space labeled as "system" it would be helpful to have a clear tmp directory button.`

I thought this means the tmp directory on the mbaas but it should probably be the tmp directory on the dyno which also is a lot easier to implement.

**Comment by wei-lee** [comment]
@pb82 yes. The customers can only see the disk space of the volume that is assigned to the container. The `system` label there means the system usage of the container, not the mbaas node. So we just need to free up as much space as we can inside the container. We don't need to worry about the mbaas nodes.

**Comment by pb82** [comment]
@wei-lee Alright, should have asked before writing this. In that case i don't need the number of nodes and temp cleanup should be a lot easier. But the other two strategies are still valid. Will send the PRs for the fh-npm cache cleanup tomorrow.

**Comment by wei-lee** [comment]
@pb82 yes, you are right. This PR is not needed, but the changes in fh-supercore, fh-dynoman and fh-dfc are all required.

**Comment by pb82** [comment]
@wei-lee Cool, thanks a lot! At least the feature will be simpler.

**Pull request closed without merging**